frontend/src/components/core/Block/Block.tsx

```diff
@@ -15,733 +15,170 @@
-        }
-
-        // We don't have any other node types!
-        throw new Error(`Unrecognized ReportNode: ${node}`)
-      }
-    )
-  }
+const BlockNodeRenderer = (props: BlockProps): ReactElement => {
+  // TODO: Move this cast into type signature of props.
+  const node = (props.node as any) as BlockNode
 
-  private isElementStale(node: ReportNode): boolean {
-    if (this.props.reportRunState === ReportRunState.RERUN_REQUESTED) {
-      // If a rerun was just requested, all of our current elements
-      // are about to become stale.
-      return true
-    }
-    if (this.props.reportRunState === ReportRunState.RUNNING) {
-      return node.reportId !== this.props.reportId
-    }
-    return false
+  if (node.isEmpty) {
+    return <></>
   }
 
-  private renderBlock(
-    node: BlockNode,
-    index: number,
-    width: number
-  ): ReactNode {
-    const BlockType = node.deltaBlock.expandable
-      ? Block.WithExpandableBlock
-      : Block
-    const enable = this.shouldComponentBeEnabled(false)
-    const isStale = this.isComponentStale(enable, node)
+  const BlockType = node.deltaBlock.expandable ? WithExpandableBlock : Block
 
-    const optionalProps = node.deltaBlock.expandable
-      ? {
-          empty: node.isEmpty,
-          isStale,
-          ...node.deltaBlock.expandable,
-        }
-      : {}
+  const enable = shouldComponentBeEnabled(false, props.reportRunState)
+  const isStale = isComponentStale(
+    enable,
+    node,
+    props.showStaleElementIndicator,
+    props.reportRunState,
+    props.reportId
+  )
 
-    const child = (
-      <BlockType
-        node={node}
-        reportId={this.props.reportId}
-        reportRunState={this.props.reportRunState}
-        showStaleElementIndicator={this.props.showStaleElementIndicator}
-        widgetMgr={this.props.widgetMgr}
-        uploadClient={this.props.uploadClient}
-        widgetsDisabled={this.props.widgetsDisabled}
-        componentRegistry={this.props.componentRegistry}
-        formsData={this.props.formsData}
-        {...optionalProps}
-      />
-    )
+  const optionalProps = node.deltaBlock.expandable
+    ? {
+        empty: node.isEmpty,
+        isStale,
+        ...node.deltaBlock.expandable,
+      }
+    : {}
 
-    if (node.deltaBlock.type === "form") {
-      const { formId, clearOnSubmit } = node.deltaBlock.form as BlockProto.Form
-      const submitButtonCount = this.props.formsData.submitButtonCount.get(
-        formId
-      )
-      const hasSubmitButton =
-        submitButtonCount !== undefined && submitButtonCount > 0
-      return (
-        <Form
-          formId={formId}
-          clearOnSubmit={clearOnSubmit}
-          width={width}
-          hasSubmitButton={hasSubmitButton}
-          reportRunState={this.props.reportRunState}
-          widgetMgr={this.props.widgetMgr}
-        >
-          {child}
-        </Form>
-      )
-    }
+  const childProps = assign({}, props, optionalProps, { node })
```

Comment:
Nit: why `assign` instead of the spread operator? `const childProps = {...props, ...optionalProps, node }` is more readable IMO

Reply:
I feel like there was a reason (also because I'm using the lodash version of `assign`???) but I can't remember right now. Changed to `...spread`.